### PR TITLE
fix(nacos): add skipped nil check for "applications" upvalue

### LINKED ISSUE #9527: bug: runtime error: /usr/local/apisix/apisix/discovery/nacos/init.lua:361: attempt to index upvalue 'applications' (a nil value)
apisix单机，etcd单机。jmeter压测时 50 thread 1 period  5 loop 执行同一个接口，报错。
image

image-20230520141905999

> maybe change value for "local applications"  is not work.
怀疑 "local applications" 刷新逻辑

### Expected Behavior

_No response_

### Error Logs

2023/05/23 16:04:22 [warn] 21022#21022: *614008 [lua] debug.lua:198: http_body_filter_phase(): call require("apisix").http_body_filter_phase() return:{} while sending to client, client: 10.202.16.131, server: _, subrequest: "/data-log/general/http2dubbo", upstream: "dubbo://10.202.68.118:20880"
2023/05/23 16:04:22 [warn] 21022#21022: *614008 [lua] debug.lua:190: http_log_phase(): call require("apisix").http_log_phase() args:{} while logging request, client: 10.202.16.131, server: _, subrequest: "/data-log/general/http2dubbo", upstream: "dubbo://10.202.68.118:20880"
2023/05/23 16:04:22 [warn] 21022#21022: *614008 [lua] plugin.lua:399: trace_plugins_info_for_debug(): Apisix-Plugins: request-id, cors while logging request, client: 10.202.16.131, server: _, subrequest: "/data-log/general/http2dubbo", upstream: "dubbo://10.202.68.118:20880"
2023/05/23 16:04:22 [warn] 21022#21022: *614008 [lua] debug.lua:198: http_log_phase(): call require("apisix").http_log_phase() return:{} while logging request, client: 10.202.16.131, server: _, subrequest: "/data-log/general/http2dubbo", upstream: "dubbo://10.202.68.118:20880"
2023/05/23 16:04:22 [error] 24067#24067: *613382 lua entry thread aborted: runtime error: /usr/local/apisix/apisix/discovery/nacos/init.lua:361: attempt to index upvalue 'applications' (a nil value)
stack traceback:
coroutine 0:
        /usr/local/apisix/apisix/discovery/nacos/init.lua: in function 'nodes'
        /usr/local/apisix/apisix/upstream.lua:259: in function 'set_upstream'
        /usr/local/apisix/apisix/init.lua:538: in function 'fun_org'
        /usr/local/apisix/apisix/debug.lua:195: in function 'http_access_phase'
        access_by_lua(nginx.conf:321):2: in main chunk, client: 10.202.16.131, server: _, request: "POST /data-log/general/http2dubbo HTTP/1.0",

### Steps to Reproduce

1. apisix one node , etcd one node & no auth
2. add route for  using nacos discovery ,like this:
{
  "type": "roundrobin",
  "scheme": "http",
  "discovery_type": "nacos",
  "discovery_args": {
    "group_name": "",
    "namespace_id": ""
  },
  "pass_host": "pass",
  "name": "data-log-dubbo",
  "service_name": "data-log",
  "keepalive_pool": {
    "idle_timeout": 60,
    "requests": 1000,
    "size": 320
  }
}
route like 
{
  "uri": "/data-log/general/http2dubbo",
  "name": "data-log_xxx",
  "methods": [    "POST"  ],
  "plugins": {
    "dubbo-proxy": {
      "method": "invoke",
      "service_name": "com.xxx.proxy.XxxService",
      "service_version": "1.0.0"
    }
  },
  "upstream_id": "data-log-dubbo",
  "status": 1
}

3. using jmeter 50/200 thread 1  5 loop 

### Environment

- APISIX version (run `apisix version`):2.15.1
- Operating system (run `uname -a`): 3.10.0-1160.80.1.el7.x86_64 #1 SMP Tue Nov 8 15:48:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`):
nginx version: openresty/1.21.4.1
built by gcc 9.3.1 20200408 (Red Hat 9.3.1-2) (GCC) 
built with OpenSSL 1.1.1s  1 Nov 2022
TLS SNI support enabled
configure arguments: --prefix=/usr/local/openresty/nginx --with-cc-opt='-O2 -DAPISIX_BASE_VER=1.21.4.1.5 -DNGX_GRPC_CLI_ENGINE_PATH=/usr/local/openresty/libgrpc_engine.so -DNGX_HTTP_GRPC_CLI_ENGINE_PATH=/usr/local/openresty/libgrpc_engine.so -DNGX_LUA_ABORT_AT_PANIC -I/usr/local/openresty/zlib/include -I/usr/local/openresty/pcre/include -I/usr/local/openresty/openssl111/include' --add-module=../ngx_devel_kit-0.3.1 --add-module=../echo-nginx-module-0.62 --add-module=../xss-nginx-module-0.06 --add-module=../ngx_coolkit-0.2 --add-module=../set-misc-nginx-module-0.33 --add-module=../form-input-nginx-module-0.12 --add-module=../encrypted-session-nginx-module-0.09 --add-module=../srcache-nginx-module-0.32 --add-module=../ngx_lua-0.10.21 --add-module=../ngx_lua_upstream-0.07 --add-module=../headers-more-nginx-module-0.33 --add-module=../array-var-nginx-module-0.05 --add-module=../memc-nginx-module-0.19 --add-module=../redis2-nginx-module-0.15 --add-module=../redis-nginx-module-0.3.9 --add-module=../ngx_stream_lua-0.0.11 --with-ld-opt='-Wl,-rpath,/usr/local/openresty/luajit/lib -Wl,-rpath,/usr/local/openresty/wasmtime-c-api/lib -L/usr/local/openresty/zlib/lib -L/usr/local/openresty/pcre/lib -L/usr/local/openresty/openssl111/lib -Wl,-rpath,/usr/local/openresty/zlib/lib:/usr/local/openresty/pcre/lib:/usr/local/openresty/openssl111/lib' --add-module=/tmp/tmp.gwSPZhqVyd/openresty-1.21.4.1/../mod_dubbo-1.0.2 --add-module=/tmp/tmp.gwSPZhqVyd/openresty-1.21.4.1/../ngx_multi_upstream_module-1.1.1 --add-module=/tmp/tmp.gwSPZhqVyd/openresty-1.21.4.1/../apisix-nginx-module-1.12.0 --add-module=/tmp/tmp.gwSPZhqVyd/openresty-1.21.4.1/../apisix-nginx-module-1.12.0/src/stream --add-module=/tmp/tmp.gwSPZhqVyd/openresty-1.21.4.1/../apisix-nginx-module-1.12.0/src/meta --add-module=/tmp/tmp.gwSPZhqVyd/openresty-1.21.4.1/../wasm-nginx-module-0.6.4 --add-module=/tmp/tmp.gwSPZhqVyd/openresty-1.21.4.1/../lua-var-nginx-module-v0.5.3 --add-module=/tmp/tmp.gwSPZhqVyd/openresty-1.21.4.1/../grpc-client-nginx-module-v0.4.0 --with-poll_module --with-pcre-jit --with-stream --with-stream_ssl_module --with-stream_ssl_preread_module --with-http_v2_module --without-mail_pop3_module --without-mail_imap_module --without-mail_smtp_module --with-http_stub_status_module --with-http_realip_module --with-http_addition_module --with-http_auth_request_module --with-http_secure_link_module --with-http_random_index_module --with-http_gzip_static_module --with-http_sub_module --with-http_dav_module --with-http_flv_module --with-http_mp4_module --with-http_gunzip_module --with-threads --with-compat --with-stream --with-http_ssl_module

- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`):

{"hostname":"TLVM202017071","id":"0880b644-4f19-49e1-af77-1be7d3d7e196","version":"2.15.1","boot_time":1684823930,"etcd_version":"3.5.0"}

- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners:
- LuaRocks version, for installation issues (run `luarocks --version`):

### PR DESCRIPTION
**Pull request by Revolyssup**
### Description

Fixes #9527  (issue)
After the 5 second wait time, it can be possible that `applications` upvalue has still not being populated and is nil. So it should be checked.
### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by AlinsRan — APPROVED** [review]

File: apisix/discovery/nacos/init.lua [hunk 1]
```diff
@@ -361,8 +361,9 @@ function _M.nodes(service_name, discovery_args)
         waiting_time = waiting_time - step
     end
 
-    if not applications or not applications[namespace_id] or not applications[namespace_id][group_name] then
-        return nil
+    if not applications or not applications[namespace_id]
+      or not applications[namespace_id][group_name] then
+      return nil
```
Comment (monkeyDluffy6017) [inline apisix/discovery/nacos/init.lua:366]: four spaces are needed

Comment (monkeyDluffy6017) [reply apisix/discovery/nacos/init.lua:366]: if not applications or not applications[namespace_id]
        or not applications[namespace_id][group_name] 
    then
        return nil

Comment (Revolyssup) [reply apisix/discovery/nacos/init.lua]: done

**Review by monkeyDluffy6017 — APPROVED** [review]

**Comment by monkeyDluffy6017** [comment]
@Revolyssup please merge the master branch

**Comment by Revolyssup** [comment]
> @Revolyssup please merge the master branch

done

**Review by Sn0rt — APPROVED** [review]

**Review by leslie-tsang — APPROVED** [review]

**Comment by lakewatcher** [comment]
我有一点疑问：仅仅增加非空判断是否依然不能解决问题。因为在高并发情况下，我从日志上看到：
1.多进程同时在发起请求和抛出错误。
2.单个进程因获取了nil而不断请求和抛出错误。那么是否会把并发的压力也传到到了服务发现上面，并且让一时的波动导致的错误让此进程一直返回错误。这是否会造成性能瓶颈。

是否应该异步刷新，增加异步刷新结果、刷新时间和空值次数的控制。
即: 1.单位时间段控制异步刷新频率【我看到配置里面的lua_shared_dict.discovery共享内存，那么请求过多也没用】。2.增加结果的时间戳每次请求只用判断时间戳发起异步刷新就行，不用影响本次请求。3.多次返回空值那就真的没发现，抛出错误是正确的逻辑。

我对lua不太了解，可能会有常识错误。如有问题请不要见怪

**Pull request merged**